### ISSUE TITLE
scrolling on an image in mobile devices

### ISSUE THREAD

**Issue by kaplanitay** [openseadragon/openseadragon#1791]
Hi all, 

I am using OpenSeaDragon on project targeted to mobile web.
I am facing an issue with the fact the user cannot scroll (moving up and down) on the viewer.

Is there a solution to this problem (something I am missing with the configuration/plug-in I can use)?

Thank you all

**Comment by iangilman** [comment]
Unfortunately I'm not aware of a comprehensive solution for this issue. There's a little discussion here: https://github.com/openseadragon/openseadragon/issues/1528 but no conclusions.

Some possibilities:

* Disable panning entirely in the viewer, which should allow the scroll event to happen, but I assume you do want panning, at least sometimes.
* Disable panning when zoomed out but enable it when zoomed in. Of course it might be tricky to communicate with the user that that's what's going on.
* Provide a little buffer on the sides of the viewer, so the user can always scroll by swiping out there.
* Disable zooming and panning entirely by default and require the user to initiate zoom/pan mode, so that otherwise they can just scroll the page as normal.
* Require two fingers for panning, leaving the single finger swipe available for scrolling. This might take some work in MouseTracker to make possible.

So I guess to some extent it comes down to what you'd like the interaction to be; then we can think about what the technical solutions for that might be.

**Comment by kaplanitay** [comment]
Hi,

Thank you very much for your respond.
I am interested in option #2. Meaning, disabling pan (and enabling scroll) when the image is zoomed out and enabling it again when the zoom in.
Can you guide me on that?

Thanks.

**Comment by iangilman** [comment]
@kaplanitay Apologies for the delayed response! The quarantine has thrown my schedule for a loop and I've had a hard time getting to OpenSeadragon things.

I've now made an example of how to do option 2: 

https://codepen.io/iangilman/pen/MWapQPB

Of course there are lots of different ways you can approach the UI.

Note that the "touch action" stuff might be better handled directly in OpenSeadragon; that could be a nice pull request to make, if you're up for it (basically including the touch action manipulation in `setMouseNavEnabled`). Anyway, it should work as an external function as well, like in my example; just not as pretty.

**Comment by superbland** [comment]
@iangilman I'm also interested in this, and wondered if there's anything I can contribute to help towards it.

The workaround we mainly use right now is to disable scroll / panning, and toggle them on if the user switches to full screen. It works and I mind this solution a lot less on desktop, but I think mobile users will miss this.

The common thing I hear from people who've raised this with me is "like google maps", which I think basically alludes to your last point:

> Require two fingers for panning, leaving the single finger swipe available for scrolling. This might take some work in MouseTracker to make possible.

I had the idea to see what would happen if `dragToPan` and `flickEnabled` on `gestureSettingsTouch` were false. It's promising that as long as zoom is enabled, two finger gestures can also pan. Unfortunately however, it looks like the events are still captured and overridden for single finger gestures.

Looking into the code a bit I noticed that [TouchUtil](https://github.com/openseadragon/openseadragon/blob/f775526d81f8d309132dc843bd1361d6d7b92d63/test/modules/events.js#L683) can report how many fingers are being used. I thought maybe that could be used in conjunction with another setting (maybe something new) [to intercept what happens during a drag event](https://github.com/openseadragon/openseadragon/blob/0d1956de050df973675e0697efb67440cfe17e70/src/viewer.js#L2961). This might be different to what you had in mind though?

**Comment by iangilman** [comment]
@superbland great! I'm imagining ultimately this would be a new thing in gesture settings so that people could turn it on or off as needed. 

Have you tried with the latest master? A lot has changed with the event handling, so it might work better. 

I believe TouchUtil is just a little testing module that simulates touches, so it wouldn't necessarily be something we could use to read about touches. I would hope it is possible somehow to get how many touches there are from MouseTracker, though. You might see if @msalsbery has any thoughts on that front. 

BTW, when you say "like Google Maps", where are you seeing this behavior? The Google Maps app has single-finger pan... are you referring to Google Maps embeds? 

Anyway, it sounds like you're on the basic right track... I think at this point it's just a matter of getting to the bottom of what's stopping the drag events getting through to the browser (and again, I recommend working off of master, since things have changed).

**Comment by superbland** [comment]
Thanks! I do have the latest master now, looks like I might have been slightly out of date before.

It doesn't look like those updates made much difference here. The `touch-action`'s that really pose a barrier the way I thought to approach this initially are the ones that we wouldn't be able to change -- on the container and canvas elements. As I understand it, those interactions need to be blocked to allow mouseTracker to do its thing, without other interactions happening in the browser. Unfortunately those browser interactions would be what I'd want to prevent the viewer swallowing up, and I can't see an obvious way to code out of that right now while the CSS is in place.

I'm still investigating at this point. I might try thinking about when a new gestureSetting is enabled, the touch-actions CSS isn't applied right at the start, but instead when a pointerDown event begins it it could apply those if there's two contacts, otherwise allow the scrolling and display the message.

I'll have a bit of a mull on this. I've had some other (vague) thoughts about maybe having something that sits above the touch-action elements that could proxy events when certain conditions are met, that could be something else to think about.

**Comment by iangilman** [comment]
Have you tried removing those `touch-action`s to see what happens? It does seem like in order for this to work we would need to control when they were on and when they were off. 

If you can figure out how to proxy that through an element on top, that would be great! I feel like I've tried that and run aground, but it was a long time ago and I don't remember the details. I think the challenging part would be convincing the browser that it was a legitimate scroll event coming through.

**Comment by daniel-keller** [comment]
I'm also looking for a solution to the "scrolling past deepzoom images" on both desktop and mobile. 

**TL;DR I can get Google map style scrolling in openseadragon v2.4.2 but not v3.0.0. See [demo](https://jsfiddle.net/f0rLgq4h/14/)**

I've seen several issues from years back dancing around this topic so forgive me if I'm repeating what's already been discussed. 

As mentioned by @superbland mimicking [Google's cooperative gesture handling](https://developers.google.com/maps/documentation/javascript/interaction#gesturehandling_cooperative) looks to be a good solution. I wanted to share a snippet of my attempt to implement this with openseadragon's input hooks. 

https://jsfiddle.net/f0rLgq4h/14/ (jsfiddle is not responsive but changing your layout setting to **Tabs (rows)** makes the demo usable on mobile)

This solution relies on two things: 
1) pinch events also allow you to pan with two fingers so for mobile support we just have to disable the default drag behavior and get those events to propagate up to the document.

2) When allowing the mobile drag event to propagate to the document I run up against the `touch-action` styles that have already been mentioned. By changing the canvas `touch-action` to `pan-y` I can get the document to scroll in openseadragon v2.4.2. This same modification applied to the canvas and container in v3.0.0 does not work.

**Comment by iangilman** [comment]
Awesome! Seems like a good start. 

Any thoughts on why the `touch-action` fix isn't working in OSD 3.0? A lot of changes were made to the event handling between versions, but we must be able to narrow it down somehow. Here's the whole diff if it's of any use: 

https://github.com/openseadragon/openseadragon/compare/v2.4.2...v3.0.0

That seems like the one thing needed here to get this code ready for use, right? 

Well, one minor detail I spotted: when you scroll on desktop over the viewer it gives you a helpful message just like on Google Maps, but it says "Ctrl + Scroll" even on Mac where it should be "Cmd + Scroll". 

At any rate, this already seems helpful to folks who need it! 

Beyond having this example code, I think this is probably a worthwhile feature to bring into OSD; plenty of people have asked for it, and it'd be nice to be able to turn it on as a viewer option. Would you be interested in making a PR with it? We'd want to remove the ViewerInputHook dependency, and there is probably some robustness work needed. If you're not interested, hopefully someone else can pick up the torch. 

Thank you for pushing this forward!

**Comment by daniel-keller** [comment]
Thanks @iangilman. I'll keep working at it in my free time. With some additional logic it will be easy to have different instructions depending on OS and event type (touch vs mouse). I'm not sure why the demo is not working in v3.0.0 but I'll look through the diff and see if anything pops out to me. 

I discovered this morning another minor issue in my solution (in that its different behavior from google maps). On a laptop's trackpad, 2 finger scroll and pinch gestures both trigger scroll events instead of a scroll and pinch event respectively. So unlike google maps, my demo requires holding down the meta/cmd key to zoom using pinch gestures as well as scroll gestures. It's something else for me to look into.

**Comment by superbland** [comment]
@daniel-keller thanks for picking this up, the demo looks promising!

I think I was probably working against the code that's giving you problems as well so I don't think I can give you much help there with the `touch-action` issue.

Where I'd gotten up to was poking around in mouseTracker. I found some interesting methods that I think play a role and was looking to investigate further, from my notes these were: `onPointerDown`, `udatePointerMove`, and `handleWheelEvent`. With any luck that might be useful for you.

After getting to that point as someone who hasn't worked with the mouseTracker before it felt like a big undertaking to try get to grips fully with its inner workings. I'd intended to come back to this when I was less busy, but sadly that hasn't happened yet :sigh:

**Comment by dtugend** [comment]
On 3.x you need to disable touch action on viewer.canvas AND viewer.container and tweak setTelementTouchAction to not use OpenSeadragon,getElement:

```
// [...]

function setElementTouchAction(element, value) {
  // remove this: // element = OpenSeadragon.getElement(element);
  if (typeof element.style.touchAction !== 'undefined') {
    element.style.touchAction = value;
  } else if (typeof element.style.msTouchAction !== 'undefined') {
    element.style.msTouchAction = value;
  }
}

// [...]

function setInteractionEnabled(value) {
  viewer.setMouseNavEnabled(value);
  setElementTouchAction(viewer.container, value ? 'none' : 'auto');
  setElementTouchAction(viewer.canvas, value ? 'none' : 'auto');
  $button.toggle(!value);
}

// [...]
```

**Comment by jonasengelmann** [comment]
I am currently tackling this issue as well. Disabling dragToPan for touch gestures already gets me pretty far. On mobile devices one can then still zoom using pinch gestures and pan around the image using two fingers. However, the only issue I am currently facing now is that one cannot scroll past the image on mobile devices. Openseadragon seems to capture all the necessary pointer events and does not propagate them. I tried to look in the code a bit, but did not really understand where this capturing logic is defined. Any ideas how this could be achieved? I basically only want Openseadragon to capture pinch, and dragging with to finger gestures, everything else should be propagated to the document. 

Using `addViewerInputHook` with `event.preventDefaultAction = true` and `event.stopHandlers = true` like in @daniel-keller's example did exactly that, however that only works with version 2.x.

**Comment by iangilman** [comment]
@msalsbery Any suggestions?

**Comment by jonw-cogapp** [comment]
Chiming in: I saw that one of my colleagues had implemented a button on top of OpenSeadragon that was a little lock/unlock icon to achieve something like the code pen example above. I believe they used a combination `mouseNavEnabled` for desktop and a `div` to block pointer events. I don't think the exact example is publicly viewable yet otherwise I'd share it here, but it might be something we can post later if that's interesting

cc. @mattc-cogapp, if you want to say more!

**Comment by mattc-cogapp** [comment]
Yeees, I added a little button that basically toggles `mouseNavEnabled` and adjusts the pointer events on the two parent divs that need it. It's fairly rudimentary, and things like pinching and zooming (i.e. gesture specific) haven't _really_ been considered in any depth, but it allows the user to scroll past a full-width viewer which was my main concern at the time.

I'm in the process of extracting it into a small plugin to put on NPM, I just need to find the time to finish it off.

**Comment by mattc-cogapp** [comment]
Yeees, I added a little button that basically toggles `mouseNavEnabled` and adjusts the pointer events on the two parent divs that need it. It's fairly rudimentary, and things like pinching and zooming (i.e. gesture specific) haven't _really_ been considered in any depth, but it allows the user to scroll past a full-width viewer which was my main concern at the time.

I'm in the process of extracting it into a small plugin to put on NPM, I just need to find the time to finish it off.
